Tools/GMSInfo/source/GMS.cpp

```diff
@@ -0,0 +1,193 @@
+
+        gms.field_C = v5;
+        gms.field_8 = *(unsigned long*)(raw + 4);
+        gms.field_14 = (*(unsigned char*)(raw + 8)) != 0;
+
+        int outBuffSize = (v5 + 15) & 0xFFFFFFF0;
+
+        auto outBuffer = std::make_unique<char[]>(outBuffSize);
+        Legacy::GMS_Decompress(&gms, outBuffer.get(), outBuffSize);
+
+        spdlog::info("??");
+        EnumerateEntities(std::move(outBuffer));
+    }
+
+    void GMS::EnumerateEntities(std::unique_ptr<char[]>&& gmsBuffer)
+    {
+        std::unique_ptr<char[]> data = std::move(gmsBuffer);
+        char* buffer = data.get();
+
+        size_t prmSize = 0;
+        size_t bufSize = 0;
+
+        auto prm = m_container->Read(m_assets->PRM, prmSize);
+        auto buf = m_container->Read(m_assets->BUF, bufSize);
+
+        if (!prm)
+        {
+            spdlog::error("GMS::EnumerateEntities() Failed to read PRM file {}", m_assets->PRM);
+            return;
+        }
+
+        if (!buf)
+        {
+            spdlog::error("GMS::EnumerateEntities() Failed to read BUF file {}", m_assets->BUF);
+            return;
+        }
+
+#define G_AT(x) (*(((int*)buffer) + x))
+        int flag_0x10 = G_AT(0x10);
+        int flag_0x11 = G_AT(0x11);
+#undef G_AT
+
+        int totalEntitiesCount = *(int*)&buffer[*(int*)buffer];
+        spdlog::info(" ---------------[ GMS ]--------------- ");
+        spdlog::info("Total entities: {}", totalEntitiesCount);
+        spdlog::info("Flag [0x10]: {:X}", flag_0x10);
+        spdlog::info("Flag [0x11]: {:X}", flag_0x11);
+
+        spdlog::info(" AF table");
+        spdlog::info("--------------");
+        int afEntitiesCount = *(int*)(&buf[flag_0x11]);
+        spdlog::info("Total entities: {}", afEntitiesCount);
+
+        int entityNum = 0;
+        char* afPtr = (char*)(&buf[flag_0x11] + sizeof(int)) + 1;
+        std::string_view afid { afPtr };
+
+        // TODO: Loop
+        spdlog::info("[{:4}] {}", entityNum, afid);
+
+        spdlog::info(" Import table");
+        spdlog::info("--------------");
+        spdlog::info("Index |    Ref    |Pointer | Type ID");
+
+        int entityLocator = 0;
+        int entityIndex = 0;
+
+        auto& db = TypesDataBase::GetInstance();
+
+        while (true)
+        {
+            entityLocator = *(int*)(&buffer[8 * entityIndex + 4] + *(int*)buffer) & 0xFFFFFF;
+            unsigned int entityType = *(int*)(&buffer[4 * entityLocator + 20]);
+            const bool isLoaderContents = entityType == Glacier::TypeId::ZLoader_Sequence_Setup_ZSTDOBJ;
+
+            int ptr = (int)&buffer[4 * entityLocator];
+            std::string typeInfoStr = db.GetEntityTypeById(entityType);
+
+            printf("#%.4d | %.9X | %.8X | %s\n", entityIndex, entityLocator, ptr, typeInfoStr.c_str());
+
+            if (db.HasDefinitionForEntityTypeId(entityType) && static_cast<Glacier::TypeId>(entityType) == Glacier::TypeId::ZHitman3_ZPlayer)
+            {
+                printf("Player!\n");
```

Comment:
Think about how to add parsers into the enumeration loop and make it cross-project.